Add solarprognose.de - Neuer PR

thlink68 opened the pull request:
Änderungsvorschläge @tolot27 eingearbeitet.

Link zu PR #19364

thlink68 committed:
Create solarprognose.de

Template um www.solarprognosede als Tarifdefinition für die Solarprognose hinzuzufügen

thlink68 committed:
Update templates/definition/tariff/solarprognose.de

Co-authored-by: andig <cpuidle@gmail.com>

thlink68 committed:
Update templates/definition/tariff/solarprognose.de

Co-authored-by: andig <cpuidle@gmail.com>

thlink68 committed:
In URI {{}} hinzugefügt

andig committed:
Update templates/definition/tariff/solarprognose.de

thlink68 committed:
Rename solarprognose.de to solarprognose-de.yaml

thlink68 committed:
Update solarprognose-de.yaml

Jeweils advanced: false entfernt weil überflüssig.

thlink68 committed:
Update templates/definition/tariff/solarprognose-de.yaml

Co-authored-by: StefanSchoof <4662023+StefanSchoof@users.noreply.github.com>

thlink68 committed:
Update templates/definition/tariff/solarprognose-de.yaml

Co-authored-by: StefanSchoof <4662023+StefanSchoof@users.noreply.github.com>

andig committed:
Update templates/definition/tariff/solarprognose-de.yaml

thlink68 committed:
Verweis auf API-Key gelöscht weil mißverständlich.

thlink68 committed:
brand auf Solarprognose akutalisieren

Co-authored-by: andig <cpuidle@gmail.com>

thlink68 committed:
template auf solarprognose aktualisieren

Co-authored-by: andig <cpuidle@gmail.com>

thlink68 committed:
Gross- und Kleinschreibung bei englischen Descriptions

thlink68 committed:
Anpassung an #19182

Co-authored-by: StefanSchoof <4662023+StefanSchoof@users.noreply.github.com>

thlink68 committed:
Update templates/definition/tariff/solarprognose-de.yaml

Co-authored-by: andig <cpuidle@gmail.com>

thlink68 committed:
Item token zu uniquetoken umbenannt

thlink68 committed:
Update solarprognose-de.yaml

Beschreibung für Item hinzugefügt.

thlink68 committed:
Update solarprognose-de.yaml

Params und render durch @tolot27 Vorschlag ersetzt.

thlink68 committed:
Rename solarprognose-de.yaml to solarprognose.yaml

thlink68 committed:
Update solarprognose.yaml

thlink68 committed:
Merge pull request #1 from thlink68/thlink68-patch-1

Update solarprognose.yaml

thlink68 committed:
Update templates/definition/tariff/solarprognose.yaml

Co-authored-by: StefanSchoof <4662023+StefanSchoof@users.noreply.github.com>

tolot27 reviewed (approved):
Sieht nun gut aus.
In einem separaten PR kann man dann noch die Implementierung von [`preferredNextApiRequestAt`](https://www.solarprognose.de/web/de/solarprediction/page/api#preferredNextApiRequestAt) vornehmen.

thlink68 committed:
Update solarprognose.yaml

nested conditional logic im URI template ersetzt durch pre-computed parameters ersetzt, damit es übersichtlicher und wartbarer wird. Außerdem Reihenfolge der Parameter angepasst.
Von @tolot27

andig commented on templates/definition/tariff/solarprognose.yaml:
```diff
@@ -0,0 +1,57 @@
+      [ .data | to_entries.[] | {
+        "start": (.key | tonumber | strftime("%FT%TZ") ),
+        "end": (.key | tonumber + 3600 | strftime("%FT%TZ")  ) ,
+        "value": (.value.[0]*1000)
```
> I've confirmed that the returned value unit is always kW.

Then remove the complexity?

tolot27 commented:
> > buildItemParams
> 
> Finden wir dafür noch einen nachvollziehbaren Namen? Sonst würde ich- da ja nciht wiederverwendet- sogar eher die alte Version stehen lassen.

Wir könnten es noch `optItemParams` oder `optionalItemParams` nennen. `item` ist ja optional. Falls es benötigt wird, muss man entweder `id` oder `token` als weiteren Parameter der URL mitgeben. Das Template `buildItemParams` zu nennen, fand ich daher zutreffend. Es kann auch gern anders heißen. Der früheren Variante einverstanden, wenn die nested item logic als letzter Parameter übergeben wird. Das verschachtelte Konstrukt ist wegen den vielen geschweiften Klammern nur schwer lesbar.

Damit die ursprüngliche Version besser verständlich ist, schlage ich folgende Variante mit Kommentar vor, wo zudem nicht mehr geprüft wird, ob `uniquetoken` einen Wert hat. Falls nicht, ist es ohnehin egal, wie die URL aussieht, da immer ein Fehler von der API zurückkommen würde. Ich glaub das ist die sinnvollste Variante.

```yaml
...
render: |
  type: custom
  tariff: solar
  forecast:
    source: http
    # Base URL with required params (format, type, algorithm, access token), then optional item-based parameters
    uri: https://www.solarprognose.de/web/solarprediction/api/v1?_format=json&type=hourly&algorithm={{ .algorithm }}&access-token={{ .token }}{{ if .item }}&item={{ .item }}{{ if .id }}&id={{ .id }}{{ else }}&token={{ unquote .uniquetoken }}{{ end }}{{ end }}
...
```

andig commented:
Perfekt!

thlink68 committed:
Update solarprognose.yaml

Wieder zurück zur früheren Variante,  die nested item logic wird als letzter Parameter übergeben. Das verschachtelte Konstrukt ist wegen den vielen geschweiften Klammern nur schwer lesbar.
Es wird zudem nicht mehr geprüft, ob uniquetoken einen Wert hat. Falls nicht, ist es ohnehin egal, wie die URL aussieht, da immer ein Fehler von der API zurückkommen würde.
